File: python/src/dsa/algorithms/insertion_sort.py
"""Insertion sort implementation.

Time: O(n^2) worst/average, O(n) best
Space: O(1)
"""

from typing import TypeVar, MutableSequence, Optional, Callable

T = TypeVar("T")


def insertion_sort(
    arr: MutableSequence[T],
    key: Optional[Callable[[T], any]] = None,
    reverse: bool = False,
) -> None:
    """Sort a mutable sequence in place using insertion sort."""
    key_fn = key or (lambda x: x)

    for i in range(1, len(arr)):
        current = arr[i]
        current_key = key_fn(current)
        j = i

        while j > 0:
            prev_key = key_fn(arr[j - 1])
            if reverse:
                should_swap = prev_key < current_key
            else:
                should_swap = prev_key > current_key

            if should_swap:
                arr[j] = arr[j - 1]
                j -= 1
            else:
                break

        arr[j] = current

File: go/algorithms/insertion_sort.go
package algorithms

import "cmp"

// InsertionSort sorts a slice in-place using insertion sort.
// Time: O(n^2) average, O(n) best (nearly sorted)
// Space: O(1)
func InsertionSort[T cmp.Ordered](arr []T) {
	for i := 1; i < len(arr); i++ {
		key := arr[i]
		j := i - 1

		for j >= 0 && arr[j] > key {
			arr[j+1] = arr[j]
			j--
		}
		arr[j+1] = key
	}
}

// InsertionSortFunc sorts a slice using a custom comparison function.
func InsertionSortFunc[T any](arr []T, less func(a, b T) bool) {
	for i := 1; i < len(arr); i++ {
		key := arr[i]
		j := i - 1

		for j >= 0 && less(key, arr[j]) {
			arr[j+1] = arr[j]
			j--
		}
		arr[j+1] = key
	}
}
